llama/generation.py

```diff
@@ -102,68 +104,124 @@ def build(
     def __init__(self, model: Transformer, tokenizer: Tokenizer):
         self.model = model
         self.tokenizer = tokenizer
+        self._generate_one_token_fn = self._generate_one_token
+        self._generate_one_token_fn = torch.compile(self._generate_one_token_fn,
+                                                    backend="torchxla_trace_once", fullgraph=True)
+
 
-    @torch.inference_mode()
+    def _generate_one_token(self, tokens, input_tokens, input_text_mask, cur_pos_tensor, 
+                            input_pos_tensor, output_pos_tensor, cache_kvs, temperature, top_p, logprobs):
+        logits, cache_kvs = self.model(input_tokens, input_pos_tensor, output_pos_tensor, cache_kvs)
+        # TODO: NEW CODE BLOCK - OPTIMIZE
+        if logprobs:
+            token_logprobs[:, prev_pos + 1 : cur_pos + 1] = -F.cross_entropy(
+                input=logits.transpose(1, 2),
+                target=tokens[:, prev_pos + 1 : cur_pos + 1],
+                reduction="none",
+                ignore_index=pad_id,
+            )
+        #####
+
+        if temperature > 0:
+            probs = torch.softmax(logits[:, -1] / temperature, dim=-1) #TODO: eval the perf impact of logits vs. logits[:,-1]
```

Comment:
Cannot simply take the last position of logits. Need to use the `output_pos_tensor`